### DISCUSSION TITLE
Debug outupt in the terminal

### DISCUSSION THREAD

**Discussion by tigawanna** [pocketbase/pocketbase#3918] [Q&A] [answered]
I noticed the serve --debug flag was removed, is there an alternative?

**Comment by ganigeorgiev** [comment]
You can set `-4` as "Min log level" in the logs settings panel:

![logs_settings](https://camo.githubusercontent.com/7cd95fc87beef7e37d212ae5adf01b1926a774fb87629e3b0098b10d7aae0b0a/68747470733a2f2f692e696d6775722e636f6d2f4d6e5a61647a582e706e67)

The same debug level messages as before are logged*, with the difference that now they can be saved and explored in the Admin UI.
_*The only thing that we don't print at the moment is the running queries, but they eventually will be added as a "Trace" log (-8)._

**Reply by tigawanna** [reply]
What about in the terminal when running ./pocketbase serve , comes in handy when debugging the hooks

**Reply by ganigeorgiev** [reply, marked as answer]
You can still use `console.*` to print to the standard output or alternatively you can add a regular debug log message as part of your application code without `$app.isDebug()` checks by just doing:

```js
// see https://pocketbase.io/docs/js-logging/
$app.logger().debug("something")
```

Logs with level below the specified settings minimum are just ignored so there shouldn't be any significant impact on the performance.

We can eventually introduce `--dev`/`--verbose` or similar flags to print in both the database and the standard output but I'm not sure whether this will be useful and will not cause confusion around the log severity levels.

**Reply by ganigeorgiev** [reply]
> comes in handy when debugging the hooks

To understand better the use case, could you elaborate on what is your usual flow when debugging the hooks?

**Reply by Plorenzo** [reply]
I also miss the sql statements on the terminal, very useful for debugging. In my case I have 2 monitors, when debugging something I usually have a terminal open with the backend running `go run main.go serve --http localhost:8090` and in the other I have the browser with the web app running, then I just click around in the app, triggering requests and seeing real time the statements on the terminal, when something is not working as expected, is quite apparent why when you see the last few statements.

**Reply by ganigeorgiev** [reply]
I agree that the queries are helpful, I'm just not sure whether it would be better to be added as a "Trace" log, just printed in the console or both. 

Will have to think a little more on this and eventually will reintroduce the old debug mode as `--dev`.

If anyone has suggestions or complains as a result of the recent logs refactoring, feel free to share your feedback and I'll consider it for the next release.

**Reply by tigawanna** [reply]
> > comes in handy when debugging the hooks
> 
> To understand better the use case, could you elaborate on what is your usual flow when debugging the hoo

It usually involves some custom SQL and it's always helpful to see what's being run , it's also helpful when viewing all the other queries and see things like how the filters translate to SQL

**Comment by whmeitzler** [comment]
I'm also looking for a more verbose console log. 
The SQL statements are very helpful for debugging. 

While I also like the slog implementation (having logs in a DB beats a file), there are cases where I'd like logged messages to the terminal as well. I often will pipe logged output through a grep call or two, and/or tee into a log file for exploration/manipulation in vscode. 

One very simple way to let the community do what it wants is to  accept a slog handler:
```
app := pocketbase.New()
	handler := slog.NewJSONHandler(os.Stdout, nil)
	app.Settings().Logs.Handler = handler 
	```
and write to that handler (if not nil) whenever you are batching logs to the DB. 

Another option could be to have `pocketbase serve` take a `logfilter` flag that prints a filtered set of logs to the terminal.

**Comment by nerdbaggy** [comment]
I would love to use `app.Logger().Info("log")` and still have it go to the console output

**Comment by ganigeorgiev** [comment]
_Just for info in case someone stumble here - I've pushed in `master` a small change with some log fixes that adds the `--dev` flag. When this flag is enabled (true by default if `go run` is used) it will print to the console the data DB SQL statements + all logs no matter of the log settings in the Admin UI. The change will be available in v0.20.1 (most likely tomorrow or the beginning of next week)._

**Reply by ganigeorgiev** [reply]
_**Edit**: the user seems to have deleted his comment at the same time when I posted mine, but I'll keep mine in case it is still not clear what `--dev` do._


`--dev` has no effect on the logs DB persistence (that's why I named it `dev` instead of the previous `debug` to avoid the confusion with the log debug levels).
It just prints whatever log and sql statements there is without debounce, no min log level checks, etc. directly to the console (Go uses stderr by default).

**Reply by dschissler** [reply]
I think that we could use a debug verbosity option or flag as well.  I would like to be able to use `app.isDev()` is the JSVM for controlling things like if views should be recompiled again on every request, and I would like to not have to marry that to receiving tons of debug verbosity.

On my RPI 4B I basically need the `app.isDev()` to be on all of the time in development or things break, and since it is on I'll be drowning in output that I can't turn off.  That could make low-tech debug fmt.Println statements hard to find, etc.

**Reply by dschissler** [reply]
Sorry, I deleted that previous post because I wrote it just after returning from the bar, and it was a bit rambling.  You immediately responded to it.

**Reply by ganigeorgiev** [reply]
> I think that we could use a debug verbosity option or flag as well.

No, it starting to become too complicated and confusing.
If you want to see only messages above some specific log levels, explore the logs that are being persisted in the Admin UI. Only the logs that satisfy your configured _"Admin UI > Logs > Logs settings > Min log level"_ setting will be persisted.

> On my RPI 4B I basically need the app.isDev() to be on all of the time in development or things break

I'm not sure that I understand what is there to break.

The `--dev` is not intended to be used during the regular/production run of your app (keep in mind that serializing and printing formatted logs has a slight performance cost too).

In short, `--dev` has similar behavior to the old `--debug` with the difference that now we print also colored `app.Logger().*` messages alongside the SQL statements.

**Reply by whmeitzler** [reply]
I appreciate @ganigeorgiev's purpose in the dev flag. Print everything to the console for a developer. I'd been holding off on upgrading to v0.20.0 because I do use the sql log in development. This solves my use case nicely.

I'd like to suggest again that pocketbase is acting both as a slog source and sink. It might be better to let a dev plug in his own slog handler into pocketbase, one of which can be the slog handler of pocketbase itself.

This way, if you need a flexible or specific logging solution, you can go and do it yourself. It also seems to flow nicely with the hooks/event source ethos of the project.

**Reply by benallfree** [reply]
Yeah SQL console logging definitely necessary for so many reasons.
